### PR TITLE
improve trace client performance using concurrent resource load

### LINKED ISSUE #5707: Improve the trace command for performance
### What problem are you facing?

We have a "facade" composition that creates ~500 resources including secrets. `crossplane beta trace` takes _minutes_ to run for this. What we really want is for this to run in _seconds_ so that we can do stuff like:

```
$ watch crossplane beta trace my-type/my-resource
```

and have it be a meaningful experience.

### How could Crossplane help solve your problem?

Make the command run faster.

So, internally in our company, we have a copy of the internal packages that is used to implement `crossplane beta trace` (`resource`, `xrm`) etc. and we have modified them to get the performance we need. We found that there are 2 parts to improving the perf of this tool:

* Support a `--concurrency` option that allows dependencies to be loaded in parallel rather than the one-at-a-time approach that is currently in place.
* Add support for controlling the QPS and burst of the k8s client used to fetch resources so that _it_ doesn't become the bottleneck for this.

I'm happy to contribute this code upstream if there is interest. 

It would be nice if, in return, I get some of these packages to not be internal such that I can write a something-like-crossplane-trace tool on top of it, without copying a bunch of code.

### If you've read so far...

you might as well read the rest.

The forked version of our command does a few more things than what is offered by `crossplane beta trace`. These are provider-specific and I imagine you _wouldn't_ want this to be in the core command. But these have been super-helpful from a practical POV and we wouldn't want to lose that. That's why I'd like the internal packages exposed for use by other similar tools.

What we do in our tool that is different is:

* For `Object` resources created by the K8s provider, we peek into the status of the manifest of the remote object to get its "true" status and messaging
* We also look for the `AsyncOperationFailure` condition that is exposed by (at least) the AWS provider so that our tests can check for a managed resource to be Synced and Ready and not have had any async operation failures.

### PR DESCRIPTION
**Pull request by gotwarlost**
### Description of your changes

fixes #5707

Add a loader with configurable concurrency to load resources in concurrent manner. The xrm client delegates to the loader for resource load and supports a functional option to set the concurrency.

Add a `--concurrency` flag for the `crank beta trace` command and configure the xrm client appropriately.

I have: 

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added or updated unit tests.
~~- [ ] Added or updated e2e tests.~~
~~- [ ] Linked a PR or a [docs tracking issue] to [document this change].~~
~~- [ ] Added `backport release-x.y` labels to auto-backport this PR.~~

### CODE REVIEW

**Comment by gotwarlost** [comment]
### A totally bad benchmark

This only serves to show that increasing concurrency _does_ matter. 

Some numbers on my laptop after hard-coding `qps` and `burst` to `50` and `100` such that the kube client doesn't backoff and become the bottleneck.

```
$ for c in 1 2 3 4 5 10 20 30 40; do echo concurrency=$c;time crank beta trace --concurrency=$c \
        our-facade/facade-1>/dev/null; done

concurrency=1

real	1m29.923s
user	0m0.835s
sys	0m0.308s

concurrency=2

real	0m45.674s
user	0m0.773s
sys	0m0.301s

concurrency=3

real	0m32.130s
user	0m0.733s
sys	0m0.298s

concurrency=4

real	0m24.895s
user	0m0.753s
sys	0m0.254s

concurrency=5

real	0m20.722s
user	0m0.739s
sys	0m0.286s

concurrency=10

real	0m20.902s
user	0m0.727s
sys	0m0.267s

concurrency=20

real	0m10.015s
user	0m0.635s
sys	0m0.217s

concurrency=30

real	0m12.343s
user	0m0.613s
sys	0m0.214s

concurrency=40

real	0m13.128s
user	0m0.598s
sys	0m0.203s
``` 

As you can see at some point, increasing the concurrency doesn't make a difference and there are diminishing returns.

**Comment by gotwarlost** [comment]
Interesting - lint is failing for go 1.22 int range usage but `make reviewable` passed on my laptop...

**Comment by gotwarlost** [comment]
Seems to me that the linter error is spurious since the code doesn't compile with that change (given 1.21 in go.mod)

**Comment by phisco** [comment]
go.mod's directive is only the minimum required version, what matters is the version of the go binary, which is set to 1.22.3 in [ci](https://github.com/crossplane/crossplane/blob/a8b579d82b7d7d3da20ce179bfe6f9a6b04df3d7/.github/workflows/ci.yml#L13).

I have to go through it more thoroughly, but in the meantime thanks for your contribution!

Just a breadcrumb: another possible optimization I left behind during the initial implementation was to use the cached discovery client, unfortunately wiring up all the usual Kubernetes flags wasn't so straightforward given that we aren't using pflag/cobra. That would probably reduce the need to fine tune the qps and burst settings.

**Comment by gotwarlost** [comment]
> what matters is the version of the go binary

not able to compile locally even with go 1.22.3

```
$ git diff
diff --git a/cmd/crank/beta/trace/internal/resource/xrm/loader.go b/cmd/crank/beta/trace/internal/resource/xrm/loader.go
index d6686f40..795a6d27 100644
--- a/cmd/crank/beta/trace/internal/resource/xrm/loader.go
+++ b/cmd/crank/beta/trace/internal/resource/xrm/loader.go
@@ -79,7 +79,7 @@ func (l *loader) load(ctx context.Context, concurrency int) {
                concurrency = defaultConcurrency
        }
        var wg sync.WaitGroup
-       for i := 0; i < concurrency; i++ {
+       for range concurrency {
                wg.Add(1)
                go func() {
                        defer wg.Done()
$ go version

go version go1.22.3 darwin/arm64

$ go install ./cmd/crank
# github.com/crossplane/crossplane/cmd/crank/beta/trace/internal/resource/xrm
cmd/crank/beta/trace/internal/resource/xrm/loader.go:82:12: cannot range over concurrency (variable of type int): requires go1.22 or later (-lang was set to go1.21; check go.mod)
```

**Comment by phisco** [comment]
🤔 you are right, I didn't realise what it was complaining about 🤔

File: cmd/crank/beta/trace/trace.go [hunk 1]
```diff
@@ -126,6 +127,11 @@ func (c *Cmd) Run(k *kong.Context, logger logging.Logger) error {
 	}
 	logger.Debug("Found kubeconfig")
 
+	// XXX: this needs to be made configurable - see TODO on line 64
+	// I used the values below for checking timing as concurrency increases
+	// kubeconfig.QPS = 50
```
Comment (jbw976) [inline cmd/crank/beta/trace/trace.go:132]: this commented out section will probably be replaced, at least for the time being until it's configurable, by https://github.com/crossplane/crossplane/pull/5737

**Review by jbw976 — COMMENTED** [review]
Definitely curious what the performance would be in your 1000+ resource scenario @gotwarlost after https://github.com/crossplane/crossplane/pull/5737 is merged. That looks to be fixing a big recent regression in performance...

**Comment by gotwarlost** [comment]
Second commit, fixes test structure and sorts children after load. Also sets QPS and burst but I will get rid of this with a rebase from master.

File: cmd/crank/beta/trace/internal/resource/xrm/loader_test.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+func countItems(root *resource.Resource) int {
+	ret := 1
+	for _, child := range root.Children {
+		ret += countItems(child)
+	}
+	return ret
+}
+
+func TestLoader(t *testing.T) {
+	tests := []struct {
```
[the remark this answers is not in the document]
Comment (gotwarlost) [reply cmd/crank/beta/trace/internal/resource/xrm/loader_test.go]: fixed in commit 497b1a3c9ed91cc0a0076700d555e98094a2084c

File: cmd/crank/beta/trace/internal/resource/xrm/loader.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+	go func() {
+		l.processing.Wait()
+		close(l.done)
+	}()
+
+	if concurrency < 1 {
+		concurrency = defaultConcurrency
+	}
+	var wg sync.WaitGroup
+	for i := 0; i < concurrency; i++ {
```
Comment (phisco) [inline cmd/crank/beta/trace/internal/resource/xrm/loader.go:83]:
```suggestion
	for range concurrency {
```

File: cmd/crank/beta/trace/internal/resource/xrm/loader_test.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+	return ret
+}
+
+func (d *simpleGenerator) getResourceChildrenRefs(_ context.Context, r *resource.Resource) []v1.ObjectReference {
+	depth := d.depthFromResource(r)
+	if depth == d.childDepth {
+		return nil
+	}
+	var ret []v1.ObjectReference
+	for i := 0; i < d.numItems; i++ {
```
Comment (phisco) [inline cmd/crank/beta/trace/internal/resource/xrm/loader_test.go:74]:
```suggestion
	ret := make([]v1.ObjectReference, 0, d.numItems)
	for range d.numItems {
```

**Comment by gotwarlost** [comment]
Not sure why the E2E tests are failing - the failing test cases seem unrelated to the changes in this PR.

**Comment by gotwarlost** [comment]
hello, how can I help moving this PR forward? Please let me know.

**Comment by gotwarlost** [comment]
Benchmarks  for 3 cases, before QPS change, after QPS change and current PR. Note that the QPS changes do **not** affect performance since access to resources are sequential with current head.

```
# before qps/ burst was set
$ git checkout 914b83a3c65e73aef00c9b96aafd57be01dbe98e
HEAD is now at 914b83a3 Merge pull request #5733 from crossplane/renovate/master-github.com-golang-jwt-jwt-v5-5.x

$ go install ./cmd/crank
$ time crank beta trace facade-type/facade-name -s >/dev/null

real	1m28.786s
user	0m1.141s
sys	0m0.385s

# after qps/ burst was set
$ git checkout b404cd0c67edd859c47df06a6e3c718d8951127d
Previous HEAD position was 914b83a3 Merge pull request #5733 from crossplane/renovate/master-github.com-golang-jwt-jwt-v5-5.x
HEAD is now at b404cd0c Merge pull request #5737 from phisco/gimme-more-requests

$ go install ./cmd/crank
$ time crank beta trace facade-type/facade-name -s >/dev/null # basically no changes, since access is one-at-a-time

real	1m29.491s
user	0m1.095s
sys	0m0.404s

# with concurrency
$ git checkout trace-perf
Previous HEAD position was b404cd0c Merge pull request #5737 from phisco/gimme-more-requests
Switched to branch 'trace-perf'
$ go install ./cmd/crank
$ for c in 1 2 4 8 16 32
> do
> echo Concurrency=$c
> time crank beta trace xeks/qa-awseuw1-cp-app-1-dc6j5 -s --concurrency=$c >/dev/null
> done 

# output slightly formatted for newlines to make the timings appear next to the concurrency value

Concurrency=1
real	1m28.942s
user	0m1.247s
sys	0m0.439s

Concurrency=2
real	0m48.002s
user	0m1.034s
sys	0m0.363s

Concurrency=4
real	0m24.246s
user	0m0.943s
sys	0m0.337s

Concurrency=8
real	0m16.333s
user	0m0.975s
sys	0m0.347s

Concurrency=16
real	0m15.118s
user	0m0.998s
sys	0m0.371s

Concurrency=32
I0617 13:25:06.604160   96865 request.go:697] Waited for 1.020916333s due to client-side throttling, not priority and fairness, request: GET:https://49528D6033D107FEF93C8D493C0EDC30.gr7.eu-west-1.eks.amazonaws.com/apis/autoscaling.aws.upbound.io/v1beta1/grouptags/qa-awseuw1-cp-app-1-es-vector-search001-b-asg-tag-local-storage

real	0m13.566s
user	0m0.986s
sys	0m0.352s

```

**Comment by jbw976** [comment]
Nice @gotwarlost, thank you for running these benchmarks in your environment, that definitely shows an improvement when running with more concurrency! 💪  Interesting to see for you that https://github.com/crossplane/crossplane/commit/b404cd0c67edd859c47df06a6e3c718d8951127d had no effect, but your explanation makes sense for environments where the RTT for each serialized network request is significant enough.

I wanted to take these changes for a test spin as well and I got different results for my environment that is just using a local kind cluster where the RTT network time from the `kubectl` client is very small.

**TL;DR:** Higher QPS/Burst (https://github.com/crossplane/crossplane/commit/b404cd0c67edd859c47df06a6e3c718d8951127d) does make a significant difference for me in this local kind scenario, while higher concurrency doesn't make much difference. I still think we should proceed with your general approach of adding concurrency support, since it clearly makes an improvement for more real world environments 🎉 , but it is still interesting to note this behavior difference 🤓 

To set up my perf testing scenario, I created a [test repo](https://github.com/jbw976/xtrace-perf) with a `XTracePerf` XRD and composition that essentially creates a variable number of `ConfigMap` objects using `provider-kubernetes` and `function-kcl`. I used the steps and resources in this repo to create a claim/XR with 250 composed resources, then ran the same steps you did to test the same commits and the changes in this PR.

`trace` perf testing repo: https://github.com/jbw976/xtrace-perf

250 `ConfigMap` objects created:
```
❯ kubectl get configmap --no-headers=true | grep '^object[-][0-9]' | wc -l
     250
```

Test https://github.com/crossplane/crossplane/commit/914b83a3c65e73aef00c9b96aafd57be01dbe98e (QPS and burst missing): **49 seconds**
```
❯ git checkout 914b83a3c65e73aef00c9b96aafd57be01dbe98e
❯ go install ./cmd/crank
❯ time crank beta trace traceperf/traceperf-tester > /dev/null
crank beta trace traceperf/traceperf-tester > /dev/null  0.42s user 0.16s system 1% cpu 49.001 total
```

Test https://github.com/crossplane/crossplane/commit/b404cd0c67edd859c47df06a6e3c718d8951127d (QPS and burst settings restored): **11.8 seconds**
```
❯ git checkout b404cd0c67edd859c47df06a6e3c718d8951127d
❯ go install ./cmd/crank
❯ time crank beta trace traceperf/traceperf-tester > /dev/null
crank beta trace traceperf/traceperf-tester > /dev/null  0.28s user 0.14s system 3% cpu 11.867 total
```

Test this PR #5726 with new concurrency setting: **Min: 11.08 seconds @ 8 concurrency, Max: 11.89 seconds @ 1 concurrency**
```
❯ git checkout gotwarlost/trace-perf
HEAD is now at 3c14da317 Apply suggestions from code review
❯ go install ./cmd/crank
❯ for c in 1 2 4 8 16 32; do echo concurrency=$c; time crank beta trace traceperf/traceperf-tester --concurrency=$c > /dev/null; done
concurrency=1
crank beta trace traceperf/traceperf-tester --concurrency=$c > /dev/null  0.32s user 0.16s system 4% cpu 11.885 total
concurrency=2
crank beta trace traceperf/traceperf-tester --concurrency=$c > /dev/null  0.31s user 0.13s system 3% cpu 11.104 total
concurrency=4
crank beta trace traceperf/traceperf-tester --concurrency=$c > /dev/null  0.29s user 0.12s system 3% cpu 11.097 total
concurrency=8
crank beta trace traceperf/traceperf-tester --concurrency=$c > /dev/null  0.28s user 0.11s system 3% cpu 11.083 total
concurrency=16
crank beta trace traceperf/traceperf-tester --concurrency=$c > /dev/null  0.31s user 0.12s system 3% cpu 11.096 total
concurrency=32
I0619 18:12:18.001407   84861 request.go:697] Waited for 1.043695958s due to client-side throttling, not priority and fairness, request: GET:https://127.0.0.1:56038/apis/kubernetes.crossplane.io/v1alpha2/objects/object-143
crank beta trace traceperf/traceperf-tester --concurrency=$c > /dev/null  0.31s user 0.13s system 3% cpu 11.114 total
```

I will take a full code review pass through the code changes as well, but I wanted to share those performance findings, as well as the reusable https://github.com/jbw976/xtrace-perf testing repo that could benefit others. Once again, I don't think these results indicate we shouldn't accept this change, as they are for a less important and less common local laptop cluster environment, with different network RTT characteristics. The concurrency support should definitely be beneficial for remote clusters 🙇‍♂️

**Comment by gotwarlost** [comment]
Those are interesting findings and certainly way different from what I have. I've noticed in the past that the network round trip time massively affects performance.

One a kind cluster where we run E2E tests, I had never noticed a performance problem (stuff loads in a few seconds) and it only when running the command on EKS clusters etc. with higher network latency that all these settings become material.

My hypothesis is that resources are loading so fast in your kind cluster that the qps/ burst of the client becomes the bottleneck because it is executing many more k8s requests per second than it would otherwise do with network latency.
In my setup where network latency is much higher, individual resources take enough time to load that it is within the default qps setting. 

I'll re-run my tests for the facade composition on a kind cluster and report here. I suspect the behavior will be closer to what you have than what I previously published.

**Comment by jbw976** [comment]
> My hypothesis is that resources are loading so fast in your kind cluster that the qps/ burst of the client becomes the bottleneck because it is executing many more k8s requests per second than it would otherwise do with network latency.
In my setup where network latency is much higher, individual resources take enough time to load that the client stays within the default qps setting.

Exactly @gotwarlost, I 💯 believe that hypothesis as well! It makes logical sense 🙇‍♂️

**Comment by gotwarlost** [comment]
Yes, I can reproduce results similar to yours for a kind cluster with ~480 resources
55s drops to 13s after the qps fix and marginal to no improvements for concurrency


```
$ git checkout 914b83a3c65e73aef00c9b96aafd57be01dbe98e
HEAD is now at 914b83a3 Merge pull request #5733 from crossplane/renovate/master-github.com-golang-jwt-jwt-v5-5.x

$ go install ./cmd/crank
$ crank beta trace facade-type/name -s

real	0m55.567s
user	0m0.609s
sys	0m0.215s

$ git checkout b404cd0c67edd859c47df06a6e3c718d8951127d
HEAD is now at b404cd0c Merge pull request #5737 from phisco/gimme-more-requests
$ go install ./cmd/crank
$ crank beta trace facade-type/name -s

real	0m12.911s
user	0m0.539s
sys	0m0.196s

$ git checkout trace-perf
Switched to branch 'trace-perf'
$ go install ./cmd/crank
$ for c in 1 2 4 8 16 32; do echo Concurrency=$c; time crank beta trace facade-type/name -s --concurrency=$c >/dev/null; done

Concurrency=1
real	0m12.847s
user	0m0.494s
sys	0m0.184s

Concurrency=2
real	0m11.714s
user	0m0.551s
sys	0m0.174s

Concurrency=4
real	0m11.585s
user	0m0.568s
sys	0m0.174s

Concurrency=8
real	0m11.409s
user	0m0.586s
sys	0m0.190s

Concurrency=16
real	0m10.978s
user	0m0.577s
sys	0m0.194s

Concurrency=32
I0620 12:00:41.315141    7075 request.go:697] Waited for 1.036958833s due to client-side throttling, not priority and fairness, request: GET:https://127.0.0.1:60368/apis/some/redacted/resource
real	0m10.040s
user	0m0.556s
sys	0m0.193s

```

**Review by jbw976 — COMMENTED** [review]
I took a pass through the code today too @gotwarlost, thanks again for this contribution! Most of my comments are around helping future readers of this code understand it quickly, i.e. simplifying things and adding some more plain english comments to help people keep track of where they are within the nested concurrency/goroutines. Let me know what you think of these suggestions, thank you! 🙇‍♂️

File: cmd/crank/beta/trace/internal/resource/xrm/loader_test.go [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+	return d.createResource(ref.APIVersion, ref.Kind, ref.Name)
+}
+
+func (d *simpleGenerator) loadResource(_ context.Context, ref *v1.ObjectReference) *resource.Resource {
+	return d.createResourceFromRef(ref)
+}
+
+func (d *simpleGenerator) depthFromResource(res *resource.Resource) int {
+	ret := 0
+	matches := reNum.FindStringSubmatch(res.Unstructured.GetName())
```
[the remark this answers is not in the document]
Comment (jbw976) [reply cmd/crank/beta/trace/internal/resource/xrm/loader_test.go:57]: remembered later on that we do need something that fulfills the `resourceLoader` interface to use in these tests, so yes an object with some mocked logic is needed instead of just pure data in table driven test cases. but if there's a way to simplify the internal model its using, that may still be nice 😇

File: cmd/crank/beta/trace/internal/resource/xrm/loader.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+// such that the resource that is loaded can be added as a child.
+type workItem struct {
+	parent *resource.Resource
+	child  v1.ObjectReference
+}
+
+// resourceLoader is a delegate that loads resources and returns child resource refs.
+type resourceLoader interface {
+	loadResource(ctx context.Context, ref *v1.ObjectReference) *resource.Resource
+	getResourceChildrenRefs(_ context.Context, r *resource.Resource) []v1.ObjectReference
```
[the remark this answers is not in the document]
Comment (gotwarlost) [reply cmd/crank/beta/trace/internal/resource/xrm/loader.go:43]: I kept it in the interface although the implementation does not need it to be symmetric with the other method. Also it should be `ctx` here and the `_` identifier in the implementation since it doesn't use it. I can remove this as well.

File: cmd/crank/beta/trace/internal/resource/xrm/loader.go [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+func (l *loader) processItem(ctx context.Context, item workItem) {
+	defer l.processing.Done()
+	res := l.l.loadResource(ctx, &item.child)
+	refs := l.l.getResourceChildrenRefs(ctx, res)
+	l.updateChild(item, res)
+	l.addRefs(res, refs)
+}
+
+// updateChild adds the supplied child resource to its parent.
+func (l *loader) updateChild(item workItem, res *resource.Resource) {
```
[the remark this answers is not in the document]
Comment (gotwarlost) [reply cmd/crank/beta/trace/internal/resource/xrm/loader.go:138]: the only reason it is so is to have locking happen at the scope of the function rather than inlining it into a bigger function.

File: cmd/crank/beta/trace/internal/resource/xrm/loader.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (gotwarlost) [reply cmd/crank/beta/trace/internal/resource/xrm/loader.go]: fixed in df75bcaaae93e7a956d44f0c62fe3a7982b70f19

File: cmd/crank/beta/trace/internal/resource/xrm/loader.go [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+	resourceLock sync.Mutex         // lock when updating the children of any resource
+	processing   sync.WaitGroup     // "counter" to track requests in flight
+	ch           chan workItem      // processing channel
+	done         chan struct{}      // done channel, signaled when all resources are loaded
+}
+
+// newLoader creates a loader for the root resource.
+func newLoader(root *resource.Resource, rl resourceLoader) *loader {
+	l := &loader{
+		l:    rl,
```
[the remark this answers is not in the document]
Comment (gotwarlost) [reply cmd/crank/beta/trace/internal/resource/xrm/loader.go]: fixed in df75bcaaae93e7a956d44f0c62fe3a7982b70f19 in the `addRefs` method since that is used from multiple places.

File: cmd/crank/beta/trace/internal/resource/xrm/loader.go [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+	go func() {
+		l.processing.Wait()
+		close(l.done)
+	}()
+
+	if concurrency < 1 {
+		concurrency = defaultConcurrency
+	}
+	var wg sync.WaitGroup
+	for range concurrency {
```
[the remark this answers is not in the document]
Comment (gotwarlost) [reply cmd/crank/beta/trace/internal/resource/xrm/loader.go:83]: resolved in df75bcaaae93e7a956d44f0c62fe3a7982b70f19

File: cmd/crank/beta/trace/internal/resource/xrm/loader.go [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+				case <-l.done:
+					return
+				case item := <-l.ch:
+					l.processItem(ctx, item)
+				}
+			}
+		}()
+	}
+	wg.Wait()
+	sortRefs(l.root)
```
[the remark this answers is not in the document]
Comment (gotwarlost) [reply cmd/crank/beta/trace/internal/resource/xrm/loader.go:101]: resolved in df75bcaaae93e7a956d44f0c62fe3a7982b70f19

File: cmd/crank/beta/trace/internal/resource/xrm/loader.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (gotwarlost) [reply cmd/crank/beta/trace/internal/resource/xrm/loader.go:146]: Did not change this since I really like locks to be function scoped. Can change it if you think that is not needed.

File: cmd/crank/beta/trace/internal/resource/xrm/loader_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (gotwarlost) [reply cmd/crank/beta/trace/internal/resource/xrm/loader_test.go]: I agree. This is ugly. Resolved in df75bcaaae93e7a956d44f0c62fe3a7982b70f19 by maintaining an explicit map of resource names to depth.

File: cmd/crank/beta/trace/internal/resource/xrm/loader.go [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/crossplane/crossplane/cmd/crank/beta/trace/internal/resource"
+)
+
+// channelCapacity is the buffer size of the processing channel, should be a high value
+// so that there is no blocking. Correctness of processing does not depend on the channel capacity.
+var channelCapacity = 1000 //nolint:gochecknoglobals // we treat this as constant only overrideable for tests.
```
[the remark this answers is not in the document]
Comment (gotwarlost) [reply cmd/crank/beta/trace/internal/resource/xrm/loader.go]: resolved in df75bcaaae93e7a956d44f0c62fe3a7982b70f19 by making it a parameter

File: cmd/crank/beta/trace/internal/resource/xrm/loader.go [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+		return
+	}
+	// ensure counters are updated synchronously
+	l.processing.Add(len(refs))
+	// free up the current processing routine even if the channel would block.
+	go func() {
+		for _, ref := range refs {
+			l.ch <- workItem{
+				parent: parent,
+				child:  ref,
```
Comment (jbw976) [inline cmd/crank/beta/trace/internal/resource/xrm/loader.go:129]: Is this susceptible to the infamous go for loop variable gotcha? https://dev.to/kkentzo/the-golang-for-loop-gotcha-1n35

Looking a bit further, it appears this was fixed with go 1.22, which we appear to be using: https://github.com/crossplane/crossplane/blob/master/go.mod#L3 

Do we completely not have to worry about this anymore? or is there any further steps we would need to take? should we continue to be defensive about this regardless, out of habit/tradition? what do you think? 🤔 
https://go.dev/blog/loopvar-preview

**Review by jbw976 — COMMENTED** [review]
Awesome @gotwarlost, thank you very much for incorporating the feedback, I appreciate you taking the time to do that. I took it for another test spin today on a live cluster and verified all the items are being returned as expected, so it is all working reliably well for me. I also ran the unit tests with `count=1000` and those passed perfectly too, so feeling pretty good about the correctness of this change.

Just one final question from me about the for loop variable gotcha - if you can confirm with your thinking that we are clear from that, then that would be good!

File: cmd/crank/beta/trace/internal/resource/xrm/loader.go (hunk 12 above)
Comment (gotwarlost) [reply cmd/crank/beta/trace/internal/resource/xrm/loader.go:129]: This code also works on Go 1.21 BTW, since there are no pointers or goroutines involved that use the loop variable outside its intended scope. Also as you rightly said Go 1.22 doesn't have this issue any more.

I can put a `ref := ref` in the loop if you think that is more consistent with the codebase, but I don't know that it is necessary.

Comment (jbw976) [reply cmd/crank/beta/trace/internal/resource/xrm/loader.go:129]: Ah yes, looking again more closely you are correct that this would work OK before go 1.22 also! A copy of the `ref` variable here is sent through the channel, so `ref` is never used outside of the scope of this loop 🙇‍♂️ 

No need to make a change here as far as I'm concerned!

**Review by jbw976 — APPROVED** [review]
Thanks again for a great contribution that will help real world usage of `crossplane beta trace` much more performant! A great addition for v1.17 🤩
